base/sort.jl

```diff
@@ -529,75 +530,99 @@ function sort!(v::AbstractVector, lo::Integer, hi::Integer, ::InsertionSortAlg,
     return v
 end
 
-# selectpivot!
+# TODO Stable doesnt make sense because rev may be true!!!!
+# return the index of the stable median of v[lo], v[mi], and v[hi]
+# that is, compare the elements and break ties using the index
 #
-# Given 3 locations in an array (lo, mi, and hi), sort v[lo], v[mi], v[hi]) and
-# choose the middle value as a pivot
+# for example, the stable median of [4, 4, 5] is the second
+# element and the stable median of [4, 4, 3] is the first element
 #
-# Upon return, the pivot is in v[lo], and v[hi] is guaranteed to be
-# greater than the pivot
-
-@inline function selectpivot!(v::AbstractVector, lo::Integer, hi::Integer, o::Ordering)
-    @inbounds begin
-        mi = midpoint(lo, hi)
-
-        # sort v[mi] <= v[lo] <= v[hi] such that the pivot is immediately in place
-        if lt(o, v[lo], v[mi])
-            v[mi], v[lo] = v[lo], v[mi]
-        end
+# TODO replace with rand(lo:hi) if Random is ever a part of Base
```

Comment:
What about hashing, instead of non-deterministic random generator?
``` julia
i1 = mod(hash(lo), hi-lo+1) + lo
i2 = mod(hash(hi), hi-lo+1) + lo
i3 = mod(hash(hi-lo), hi-lo+1) + lo
```

Reply:
Good idea!

Benchmarking with
```julia
@btime sort!(x, alg=QuickSort) setup=(x=rand(1:100000, 2000)) evals=1 samples=50000;
```
I get
`42.216 μs (1.00)` for median of `lo`, `mi`, `hi`
`45.960 μs (1.08)` for `rand`
`48.168 μs (1.14)` for median of 3 hashes
`44.711 μs (1.06)` for median of `lo`, `hi`, and `mod(hash(hi), hi-lo-1) + lo + 1`.
But these differences are pretty close to my limit for benchmark accuracy, so I don't know if they are accurate.

I'm still looking into worst-case behavior now.

Reply:
Worst cases can be generated quite easily, as in [this example](https://gist.github.com/petvana/8008a1ed4a2d5513f4bb2e398978b695). The hashing cannot eliminate them; however, it seems less probable that any algorithm would produce them accidentally. What I see more interesting is whether such pathological cases can be detected by the number of `while` iterations. It would enable us to call some fallback algorithm, like merge sort, or find more accurate median.

Reply:
In this PR, it is even easier to find a quadratic case: `vcat(1:10000, 10000:-1:1)` and `vcat(10000:-1:1, 1:10000)` are both quadratic there but efficient on master because the process of selecting a pivot before the partition and then swapping the pivot in after disrupt the order. That swapping around is unstable. To keep stability and not decrease robustness we must either randomize or hash the majority of the indices used to select pivots (or add worst-case detection and a fallback). Viable options include: 
.1) insert 
```julia
lo = mod(hash(lo), mi-lo) + lo
hi = mod(hash(hi), hi-mi) + mi + 1
```
at the top of the `select_pivot` function to hash 2/3 of the indices.
2) hash all three indices independently as you do above
3) replace the whole select pivot function with `mod(hash(lo), hi-lo+1) + lo` to hash 1/1 of the indices
4) detect worst-case and dispatch to a more robust algorithm

I benchmarked this test function on the current version and options 1-3
```
function test()
    @btime sort!(x, alg=QuickSort) setup=(x=rand(1:100000, 2000)) evals=1 samples=50000;
    @btime sort!(x, alg=QuickSort) setup=(x=sort!(rand(1:100000, 2000))) evals=1 samples=50000;
    @btime sort!(x, alg=QuickSort) setup=(x=vcat(1:10000,10000:-1:1)) evals=1 samples=3000;
    @btime sort!(x, alg=QuickSort) setup=(x=vcat(10000:-1:1, 1:10000)) evals=1 samples=3000;
    nothing
end
```
With these results:
```
As is (1a30832b604f7f153ba3965f799a6f2924e98be0)
  42.039 μs (1 allocation: 15.75 KiB)
  19.532 μs (1 allocation: 15.75 KiB)
  100.561 ms (2 allocations: 156.30 KiB)
  97.124 ms (2 allocations: 156.30 KiB)
1)
  46.403 μs (1 allocation: 15.75 KiB)
  23.195 μs (1 allocation: 15.75 KiB)
  501.198 μs (2 allocations: 156.30 KiB)
  505.714 μs (2 allocations: 156.30 KiB)
2)
  47.160 μs (1 allocation: 15.75 KiB)
  26.952 μs (1 allocation: 15.75 KiB)
  469.629 μs (2 allocations: 156.30 KiB)
  455.273 μs (2 allocations: 156.30 KiB)
3)
  47.103 μs (1 allocation: 15.75 KiB)
  28.547 μs (1 allocation: 15.75 KiB)
  491.788 μs (2 allocations: 156.30 KiB)
  491.719 μs (2 allocations: 156.30 KiB)
```
I favor option 3 because it is the simplest and I don't value the 22% speedup in the presorted case particularly highly because there is already a presorted check in the pipeline.

There are still quadratic worst cases with a hash-chosen pivot, but I think they are similarly unlikely to emerge than the quadratic cases with master's version. It is certainly possible to detect and eliminate quadratic runtime. However, I'd like to call that out of scope for this PR so long as there isn't a regression.

Reply:
I think that using a random pivot is strongly worth considering since it prevents the possibility of denial of service attacks.

Reply:
Agreed. My preference right now is to move all the sorting algorithms in Sort.jl except for InsertionSort to a stdlib so that they can depend on all of Base and on other stdlibs. Then we would replace `mod(hash(lo), hi-lo+1) + lo` with `rand(lo:hi)`. I think there are other cases where fast sorting algorithms benefit from having more dependencies, and I doubt that we will see much if any regression in build times with only insertion sort.

Reply:
Right now we can't use `rand` in Sort.jl because `rand` is defined in a stdlib.

Reply:
`rand` is also not cryptographically secure, so to get full protection against a strong adversary we would still need quadratic case detection.

Reply:
Update: pivot selection is randomized!

Reply:
It may sound too conservative, but wouldn't it be better to detect the quadratic cases and thus abandon randomization (given the (un)expected side effects)?

Reply:
Perhaps. Randomization proved harder to implement than I anticipated. However now that it's done, looking back over the total diff, it's pretty small. To implement it takes 2 lines of code and to accommodate it elsewhere, we keep using mergesort in `names` and manually initialize the rng state in one contrib/ script.

With quadratic case detection, the algorithm becomes a bit more complicated. The existing median-of-three pivot selection system swaps some elements as it selects them which breaks stability. Without the swapping, though, inceraing then decreasing is quadratic, and even with quadratic detection, I expect a detected quadratic case will be much slower than randomized. While quadratic case detection is able to eliminate quadratic runtime in all cases, it does not eliminate pretty bad runtime in more common problematic cases.

Quadratic case detection | Randomized pivot selection
--|--
Impossible to have quadratic runtime | Vanishingly unlikely to have quadratic runtime or even substantially worse than average runtime so long as global rng is not compromised
Does not mutate global random state | Simpler algorithm (fewer parameters to tune, fewer branches to test, fewer lines of code)
Consistent runtime for a given input (helps with benchmarking) | There is no input order with a substantially above average expected runtime
| | Already implemented
| | Supports the possibility of eventually (2.0?) moving MergeSort to SortingAlgorithms.jl

Reply:
imo, randomized is much better.

Reply:
imo, randomized is slightly better because simplicity is important to me.
Any factors you find especially important in judging it much better?